src/module.c

```diff
@@ -5984,10 +6010,80 @@ RedisModuleString *RM_CreateStringFromCallReply(RedisModuleCallReply *reply) {
 }
 
 /* Modifies the user that RM_Call will use (e.g. for ACL checks) */
-void RM_SetContextUser(RedisModuleCtx *ctx, const RedisModuleUser *user) {
+void RM_SetContextUser(RedisModuleCtx *ctx, const RedisModuleUser *user)
+{
     ctx->user = user;
 }
 
+/* Returns an object that provides a client that can be attached to a
+ * RedisModuleContet object to provide a persistent client usage
+ * between multiple RM_Calls.  This simulates a network connected
+ * client and enables modules to use WATCH/MULTI/Exec transaction
+ */
+RedisModuleClient *RM_CreateModuleClient(RedisModuleCtx *ctx)
+{
+    UNUSED(ctx);
+
+    client *c = moduleAllocTempClient(NULL);
+    c->flags |= CLIENT_MODULE_PERSISTENT;
+
+    RedisModuleClient *rmc = zmalloc(sizeof(RedisModuleClient));
+    rmc->client = c;
+
+    return rmc;
+}
+
+int RM_FreeModuleClient(RedisModuleCtx *ctx, RedisModuleClient *client)
+{
+    UNUSED(ctx);
+
+    if (!client) return REDISMODULE_OK;
+
+    if (client->client->flags & CLIENT_BLOCKED) {
+        return REDISMODULE_ERR;
+    }
+
+    /* remove CLIENT_MODULE_PERSISTENT -> client to be returned to pool */
+    client->client->flags &= ~CLIENT_MODULE_PERSISTENT;
+    moduleReleaseTempClient(client->client);
+    zfree(client);
+
+    return REDISMODULE_OK;
+}
+
+/* Enables persistent clients to work with ACLs.  Requires using RM_Call with
+ * the 'C' flag to check ACLs
+ */
+void RM_SetClientUser(RedisModuleClient *client, RedisModuleUser *user)
+{
+    if (user) {
+        client->client->user = user->user;
+    } else {
+        client->client->user = NULL;
+    }
+}
+
+/* Sets the persistent client object on the given context.  To unset the
+ * client, call with the client seto NULL)
+ */
+void RM_SetContextClient(RedisModuleCtx *ctx, RedisModuleClient *client)
+{
+    ctx->persistent_client = client;
+}
+
+/* Get the client flags that are exported to Redis Modules */
+uint64_t RM_GetClientFlags(RedisModuleClient *client)
```

Comment:
maybe we better also add an API similar to RM_GetContextFlagsAll, so that user using a new header and and old server, can detect what's expected to be supported.

Reply:
done